### PR TITLE
Add `Format` to the Scala CAPI client

### PR DESCRIPTION
**Pull request by buck06191**
## What does this change?

Rendering platforms have moved to use the `Format` type to describe the
format that rendered data should take. This includes three specific
fields:
 - `Display`
 - `Design`
 - `Theme (Pillar/Special)`

Currently the platforms handle the decision logic around how to
construct this object on platform but ideally this will be done upstream
in order to standardise across platforms and to simplify the parsing of
data.

Here  I upstream the `Format` type to the CAPI Scala client.

### Changes:
- Introduce the `Display`, `Theme` and `Design` traits and their
associated `case object`s.
- Refactor `CAPIModelEnrichment` to move code shared by `RichContent`
and `Format` to the upper scope.
- Reproduce the decision logic used by platforms to construct the
`Format` object.

## How to test

I've added unit tests for this and they can be run using whatever testing process you prefer e.g. `sbt test`

## How can we measure success?

Improved DevX for anyone working with the `Format` type, and ideally tidy refactors of the platform code to use this instead of multiple checks and flags.

## Have we considered potential risks?

I don't think there are any risks. This doesn't affect the raw data and shouldn't remove backwards compatibility.

### CODE REVIEW

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:62]: This is the original `DesignType` for these so I decided not to touch them. However they're handled down the line, the current approach is to make a `deadBlog` an `Article`. We'll be passing it through asa different `Design` for the new `Format` work though

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:120]: > I was having this discussion with @nicl and @gtrufitt earlier, is it better to use a more descriptive name instead of 'c'? I'm actually confused here what 'c' represents! 

Happy to change this. I was following the convention from the existing code, but I agree that we can make it simpler. `c` is something of the `Content` type, and overall this `val` is a `ContentFilter` which is a function mapping from `Content` to a `Boolean`.

> The isComment(c) but might be a problem in that we need to both check designtype and pillar in combination here
I can maybe make this clearer in the code but this is actually doing that. `isComment(c)` checks if it's of design type comment, and then `isPillar("News")(c)` checks the pillar.

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:125]: I mapped these directly from `@guardian/types`. Is there anything in particular for `Arts`? Or have I just missed one?

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:120]: I suspect it's more that I just can't read Scala that well but I can imagine this code being something that is read a lot in future by people trying to understand how the model works at the Guardian and a lot of them won't be Scala devs either so the clearer the better I guess.

Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:120]: Another good place to explain to people how a pice of code works is with the tests. It might be better to focus there and let the code be what it needs to be?

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:125]: I _think_ `Arts` is what comes out of CAPI. It gets mapped to `Culture` later in Frontend. For reasons. If you search the codebase for 'arts' you see a few cases like [this one](https://github.com/guardian/frontend/blob/e71dc1c521672b28399811c59331e0c2c713bf00/onward/app/models/dotcomponents/DotcomponentsOnwardsModels.scala#L41)

Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:125]: [Annnd](https://github.com/guardian/frontend/blob/d7d261ac608a5b1c27a5977a5ef70b978cce91c4/static/src/javascripts.flow.archive/projects/common/modules/audio/index.js#L80) `pillar = 'books'` 😱  I've never heard of this, I'm guessing this is super legacy...

Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:125]: That's great. I'll update this to reflect that then

**Comment by shtukas** [comment]
Looks good +1

**Comment by tonytw1** [comment]
This textbook rejection of a PR like this is "putting domain logic like this in a language specific API client is bad".

But this is been extensively reviewed by the submitting teams and there motive is deduplication of code which is already in CAPI consumers.
This PR does the hard work of getting the different consumers to agree what their domain model should like like.
We should encourage this and offer to upstream these new domain model from the client into CAPI.

CAPI should approve this and we can roll forward to get it into the ideal place.

**Comment by tonytw1** [comment]
Furthermore.... This will be a versioned release so if the teams submitting this PR can consume it immediately they will be doing the regression testing. Other consumers can stay on or revert back last good release if this causes issues.

**Review by tonytw1 — APPROVED** [review]

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (gtrufitt) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:32]: We'll need https://github.com/guardian/types/pull/83 to match up with Types

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (gtrufitt) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:85]: I think we are missing `Letter` Design? https://github.com/guardian/types/blob/main/src/format.ts#L24

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (gtrufitt) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:90]: Are we also missing `Interactive` design? https://github.com/guardian/types/blob/main/src/format.ts#L32

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (gtrufitt) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:191]: A note here that we need to update https://github.com/guardian/types/blob/main/src/format.ts#L37-L43 to remove `Column` Display.

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:85]: That's a good question. We don't currently have `Letter` as a `Design` in our [`decideDesign.ts`](https://github.com/guardian/dotcom-rendering/blob/main/src/web/lib/decideDesign.ts) function.
@oliverlloyd do you know if whether `@guardian/types` is correct or if DCR is?

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:90]: I don't think we technically support an `Interactive` design on DCR? It's not in our [`decideDesign`](https://github.com/guardian/dotcom-rendering/blob/main/src/web/lib/decideDesign.ts) function at least. If it does need supporting, I can add it in here and that way we'll just need to add it to the `decideDesign` function to translate it from the Scala JSON output to whatever Typescript needs.

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:85]: `@guardian/types` trumps DCR. We haven't yet updated the version of the /types package we're using but when we do we will add support for `Letter`

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:90]: Should the be an `Interactive` `Design` type? I'm not sure what it represents? @JamieB-gu is there a need for this in Apps or could we remove it?

Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:90]: It's in the [apps-rendering decision functions](https://github.com/guardian/apps-rendering/blob/1062b75faf0848cb6ae2bf4fd894b9a320352ce1/src/item.ts#L286-L290), so I'll add it in. If it's in here and then removed, it's not a difficult thing to rip out (which is part of the motivation for this work in the first place). 

I think that the work here should be done with a focus on targeting a specific version of `@guardian/types`, and at this point this is ever so slightly ahead of the current version by having different `Design`s for dead and live blogs, as well as removing the `Column` display type. But after this initial step of aligning the two, the goal should be to update the two in parallel, with references to the appropriate version of each within the changelog

**Review by JustinPinner — APPROVED** [review]
I agree with the commentary where we acknowledge that this isn't the ideal resting place for this functionality, and I have no objections to using it as a temporary measure prior to relocating into tools and CAPI. It looks like a well enough isolated change and test coverage looks good too. I can't comment on the intricacies of relationships between types / DCR / dotcom etc, so I'll approve on the basis that you're all happy that everything is covered before you merge 👍

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala]: An alternative to changing the variable name would be to add a type annotation. So:

```scala
def isPillar(pillar: String): ContentFilter = (c: Content) => c.pillarName.contains(pillar)
```

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala]: Isn't "arts" the legacy name for "culture"? As in, it got renamed at some point?

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:87]: Letter is a recent addition: https://github.com/guardian/types/pull/69

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:92]: `Design.Interactive` is at the moment [derived from](https://github.com/guardian/apps-rendering/blob/c671b3e4d0a23a8adbe3ad93443076b0f7894fc7/src/capi.ts#L40) `ContentType.INTERACTIVE`. I believe it's there so that we can build the empty placeholder articles for full-page interactives (`ng-interactive`/`type/interactive` - example [here](https://www.theguardian.com/environment/ng-interactive/2020/oct/05/climate-data-dashboard-carbon-atmosphere-sea-level-arctic-ice)). Note that we explicitly render these [without any styles](https://github.com/guardian/apps-rendering/blob/986336e5970154754f4f9a294961a4e17569667b/src/components/body.tsx#L56) because the interactive code takes over.

I'm curious as to how DCR models these kinds of articles if it doesn't use the `Design` 🤔?

**Comment by JamieB-gu** [comment]
> We should encourage this and offer to upstream these new domain model from the client into CAPI.

> temporary measure prior to relocating into tools and CAPI

Does this mean moving it into the Thrift definitions?

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:24]: [AR also has](https://github.com/guardian/apps-rendering/blob/bd7214635ea43813673e2fb40db783e45069bd58/src/item.ts#L237) `type/picture` in this category, see https://github.com/guardian/apps-rendering/pull/1146.

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:76]: Did you want to reuse your `displayHintExistsWithName` function here?

```scala
val isPhotoEssay: ContentFilter = content => displayHintExistsWithName("photoessay")(content)
```

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:94]: @oliverlloyd are we keeping Photo Essays as a separate `Design`, or is this temporary until we subsume MultiImage into a normal body element?

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (JamieB-gu) [inline client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:124]: What happens for special reports assigned in the targeting tool?

**Review by JamieB-gu — COMMENTED** [review]
Really nice! Very pleased that this is starting to make its way upstream 😀

**Comment by buck06191** [comment]
> > We should encourage this and offer to upstream these new domain model from the client into CAPI.
> 
> > temporary measure prior to relocating into tools and CAPI
> 
> Does this mean moving it into the Thrift definitions?

It should do, yes. I guess we (platforms) all need to have a discussion as to how we do this, then standardise on it and move forwards

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:124]: Hmm, I did this off the back of different decision functions I found across the different codebases. If there's one missing, then we'll need to implement it in a future version. I imagine the process of integrating this all into the platforms will highlight missing areas and then we can update this to reflect that .

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (buck06191) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:24]: Thanks for this. I'll add it to a card for changes to make on a future version

File: client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala
Comment (oliverlloyd) [reply client/src/main/scala/com.gu.contentapi.client/utils/CapiModelEnrichment.scala:94]: There is still at least one design aspects of Photo Essays that is unique: coloured captions. So unless we lose that it's here to stay

**Pull request merged**
